### PR TITLE
Add `flatten` to Source[T]

### PR DESCRIPTION
**Pull request by nimatrueway**
To address https://github.com/softwaremill/ox/issues/197

### CODE REVIEW

File: core/src/main/scala/ox/channels/SourceOps.scala [hunk 1]
```diff
@@ -393,6 +393,39 @@ trait SourceOps[+T] { outer: Source[T] =>
...
+  /** Pipes the elements of child sources into the output source. If the parent source or any of the child sources emit an error, the
+    * pulling stops and the output source emits the error.
+    */
+  def flatten[U](using Ox, StageCapacity, T <:< Source[U]): Source[U] = {
+    val c2 = StageCapacity.newChannel[U]
+
+    forkPropagate(c2) {
+      var pool = List[Source[T] | Source[U]](this)
+      repeatWhile {
+        selectOrClosed(pool) match {
```
Comment (nimatrueway) [inline core/src/main/scala/ox/channels/SourceOps.scala:405]: The fact that we can not know the result of select stems from which select statement is limiting. In Golang there's a hacky way to do it.
https://stackoverflow.com/a/19992525/1556045

I think both `(v: V)` or `(c: ChannelClosed)` results if they hint that which select/source they stem from would make the API more robust.

**Comment by adamw** [comment]
Thanks for the PR - the only reason such a method didn't exist before is because nobody wrote it :). There are some test failures, though.

As for adding a select variant which would also return the clause which was completed - maybe you can create a separate issue for that?

**Comment by nimatrueway** [comment]
> Thanks for the PR - the only reason such a method didn't exist before is because nobody wrote it :). There are some test failures, though.
> 
> As for adding a select variant which would also return the clause which was completed - maybe you can create a separate issue for that?

Thanks for taking a look @adamw, I created a new ticket and made the test pass. Let me know if there's any concern with the readability of the tests or more test cases would help.

**Comment by adamw** [comment]
The test are great, thanks! :)

**Pull request merged**